File: 19-RML/Classwork/Class_03/script_classificacao.R
# ==================================================
# 1. Carregar pacotes
# ==================================================

# Instala pacotes se ainda não estiverem instalados
install.packages("tidymodels", dependencies = TRUE)
install.packages("tidyverse", dependencies = TRUE)
install.packages("ranger", dependencies = TRUE)
install.packages("xgboost", dependencies = TRUE)
install.packages("dlookr", dependencies = TRUE)
install.packages("corrplot", dependencies = TRUE)
install.packages("kernlab", dependencies = TRUE)

# Carregando pacotes
library(tidymodels)   # traz recipes, parsnip, workflows, tune, yardstick, dials, rsample
library(tidyverse)    # manipulação e visualização
library(ranger)       # Random Forest
library(xgboost)      # XGBoost
library(dlookr)       # estatísticas descritivas
library(corrplot)     # gráfico de correlação
library(kernlab)      # Engine para SVM


# ==================================================
# 2. Importar e Preparar a base de dados
# ==================================================

df = read_csv("aula_02_exemplo_02.csv")
df$custo_medico = factor(df$custo_medico,levels = c("Alto_Custo", "Baixo_Custo"))


# Visualizar as primeiras linhas
head(df)
nrow(df)

# ==================================================
# 4. Análise Exploratória
# ==================================================
diagnose(df)
diagnose_numeric(df)
diagnose_category(df)

# Gráfico de barras para a nova variável alvo
df %>%
  ggplot(aes(x = custo_medico, fill = custo_medico)) +
  geom_bar() +
  labs(title = "Distribuição de Custo Médico", x = "Categoria de Custo", y = "Contagem")

# Relação entre ser fumante e o custo médico
df %>%
  ggplot(aes(x = smoker, fill = custo_medico)) +
  geom_bar(position = "fill") +
  labs(title = "Proporção de Custo Médico por Hábito de Fumar",
       x = "Fumante", y = "Proporção")


# ==================================================
# 5. Correlações numéricas
# ==================================================
correlacao = df %>% 
  select(age, bmi, children) %>%
  cor(method = "spearman")

corrplot(correlacao, method = "color", type = "upper",
         addCoef.col = "black", tl.col = "black", tl.srt = 45)


# ==================================================
# 6. Divisão dos Dados
# ==================================================

# Dividir 80% treino / 20% teste, estratificando pela variável alvo
set.seed(123)
split = initial_split(df, prop = 0.8, strata = custo_medico)
train_data = training(split)
test_data  = testing(split)

# ==================================================
# 7. Pré-processamento e Transformação dos dados (Receita)
# ==================================================

# A receita agora prevê `custo_medico`
rec = recipe(custo_medico ~ ., data = train_data) %>%
  step_impute_median(all_numeric_predictors()) %>%
  step_corr(all_numeric_predictors(), threshold = 0.9) %>%
  step_dummy(all_nominal_predictors()) %>%
  step_normalize(all_numeric_predictors())

# ==================================================
# 8. Definição dos Modelos e Tunagem de Hiperparâmetros
# ==================================================

# --- Definição dos Modelos de CLASSIFICAÇÃO ---

# 1. Regressão Logística (com regularização)
log_model = logistic_reg(penalty = tune(), mixture = 1) %>% # mixture = 1 é Lasso (L1)
  set_engine("glmnet") %>%
  set_mode("classification")

# 2. Random Forest
rf_model = rand_forest(
  mtry = tune(),
  min_n = tune(),
  trees = 500
) %>%
  set_engine("ranger") %>%
  set_mode("classification")

# 3. XGBoost
xgb_model = boost_tree(
  trees = 1000,
  tree_depth = tune(),
  learn_rate = tune(),
  loss_reduction = tune()
) %>%
  set_engine("xgboost") %>%
  set_mode("classification")

# --- Workflows: Combina receita e modelo ---
log_wf = workflow() %>% add_recipe(rec) %>% add_model(log_model)
rf_wf  = workflow() %>% add_recipe(rec) %>% add_model(rf_model)
xgb_wf = workflow() %>% add_recipe(rec) %>% add_model(xgb_model)
# --- Validação Cruzada ---
set.seed(123)
cv_folds = vfold_cv(train_data, v = 5, strata = custo_medico)

# --- Grids para Tuning ---
set.seed(123)
log_grid = grid_random(penalty(), size = 10)
rf_grid = grid_random(mtry(range = c(1, 8)), min_n(range = c(2, 10)), size = 10)
xgb_grid = grid_random(tree_depth(range = c(2, 10)), learn_rate(range = c(0.01, 0.3)), loss_reduction(), size = 10)

# --- Métricas de Classificação ---
class_metrics = metric_set(roc_auc, precision,recall)

# --- Ajuste dos modelos com validação cruzada (Tuning) ---
log_tune = tune_grid(log_wf, resamples = cv_folds, grid = log_grid, metrics = class_metrics)
rf_tune  = tune_grid(rf_wf, resamples = cv_folds, grid = rf_grid, metrics = class_metrics)
xgb_tune = tune_grid(xgb_wf, resamples = cv_folds, grid = xgb_grid, metrics = class_metrics)

# --- Seleção dos melhores hiperparâmetros (baseado em roc_auc) ---
best_log = select_best(log_tune, metric = "roc_auc")
best_rf  = select_best(rf_tune, metric = "roc_auc")
best_xgb = select_best(xgb_tune, metric = "roc_auc")

# --- Finaliza workflows com melhores parâmetros ---
final_log_wf = finalize_workflow(log_wf, best_log)
final_rf_wf  = finalize_workflow(rf_wf, best_rf)
final_xgb_wf = finalize_workflow(xgb_wf, best_xgb)

# --- Ajusta os modelos finais nos dados de treino ---
final_log_fit = fit(final_log_wf, data = train_data)
final_rf_fit  = fit(final_rf_wf, data = train_data)
final_xgb_fit = fit(final_xgb_wf, data = train_data)

# ==================================================
# 9. Previsões Conjunto de Treino e Teste
# ==================================================

# Previsões - Regressão Logística (Treino)
pred_log_treino =predict(final_log_fit, new_data = train_data, type = "prob") %>%
  bind_cols(train_data)

# Previsões - Random Forest (Treino)
pred_rf_treino = predict(final_rf_fit, new_data = train_data, type = "prob") %>%
  bind_cols(train_data)

# Previsões - XGBoost (Treino)
pred_xgb_treino = predict(final_xgb_fit, new_data = train_data,type = "prob") %>%
  bind_cols(train_data)



# --- Previsões no conjunto de TESTE ---

# Previsões - Regressão Logística (Teste)
pred_log_teste = predict(final_log_fit, new_data = test_data,type = "prob") %>%
  bind_cols(test_data)

# Previsões - Random Forest (Teste)
pred_rf_teste = predict(final_rf_fit, new_data = test_data,type = "prob") %>%
  bind_cols(test_data)

# Previsões - XGBoost (Teste)
pred_xgb_teste = predict(final_xgb_fit, new_data = test_data,type = "prob") %>%
  bind_cols(test_data)


# ==================================================
# 10. Analise de limiar com os dados de treino
# ==================================================


# --- Aplicando limiares aos dataframes de TREINO ---


# Regressão Logística - Treino
pred_log_treino = pred_log_treino %>%
  mutate(
    .pred_class_0.1 = factor(ifelse(.pred_Alto_Custo >= 0.1, "Alto_Custo", "Baixo_Custo"),levels = c("Alto_Custo", "Baixo_Custo")),
    .pred_class_0.5 = factor(ifelse(.pred_Alto_Custo >= 0.5,  "Alto_Custo", "Baixo_Custo"),levels = c("Alto_Custo", "Baixo_Custo")),
    .pred_class_0.7 = factor(ifelse(.pred_Alto_Custo >= 0.7,  "Alto_Custo", "Baixo_Custo"),levels = c("Alto_Custo", "Baixo_Custo")),
    .pred_class_0.9 = factor(ifelse(.pred_Alto_Custo >= 0.9,  "Alto_Custo", "Baixo_Custo"),levels = c("Alto_Custo", "Baixo_Custo"))
  )

# Random Forest - Treino
pred_rf_treino = pred_rf_treino %>%
  mutate(
    .pred_class_0.1 = factor(ifelse(.pred_Alto_Custo >= 0.1, "Alto_Custo", "Baixo_Custo"),levels = c("Alto_Custo", "Baixo_Custo")),
    .pred_class_0.5 = factor(ifelse(.pred_Alto_Custo >= 0.5,  "Alto_Custo", "Baixo_Custo"),levels = c("Alto_Custo", "Baixo_Custo")),
    .pred_class_0.7 = factor(ifelse(.pred_Alto_Custo >= 0.7,  "Alto_Custo", "Baixo_Custo"),levels = c("Alto_Custo", "Baixo_Custo")),
    .pred_class_0.9 = factor(ifelse(.pred_Alto_Custo >= 0.9, "Alto_Custo", "Baixo_Custo"),levels = c("Alto_Custo", "Baixo_Custo"))
  )

# XGBoost - Treino
pred_xgb_treino = pred_xgb_treino %>%
  mutate(
    .pred_class_0.1 = factor(ifelse(.pred_Alto_Custo >= 0.1,  "Alto_Custo", "Baixo_Custo"),levels = c("Alto_Custo", "Baixo_Custo")),
    .pred_class_0.5 = factor(ifelse(.pred_Alto_Custo >= 0.5,  "Alto_Custo", "Baixo_Custo"),levels = c("Alto_Custo", "Baixo_Custo")),
    .pred_class_0.7 = factor(ifelse(.pred_Alto_Custo >= 0.7,  "Alto_Custo", "Baixo_Custo"), levels = c("Alto_Custo", "Baixo_Custo")),
    .pred_class_0.9 = factor(ifelse(.pred_Alto_Custo >= 0.9,  "Alto_Custo", "Baixo_Custo"), levels = c("Alto_Custo", "Baixo_Custo"))
  )

# --- Aplicando limiares aos dataframes de TESTE ---

# Regressão Logística - Teste
pred_log_teste = pred_log_teste %>%
  mutate(
    .pred_class_0.1 = factor(ifelse(.pred_Alto_Custo >= 0.1,  "Alto_Custo", "Baixo_Custo"), levels =  c("Alto_Custo", "Baixo_Custo")),
    .pred_class_0.5 = factor(ifelse(.pred_Alto_Custo >= 0.5,  "Alto_Custo", "Baixo_Custo"), levels =  c("Alto_Custo", "Baixo_Custo")),
    .pred_class_0.7 = factor(ifelse(.pred_Alto_Custo >= 0.7,  "Alto_Custo", "Baixo_Custo"), levels =  c("Alto_Custo", "Baixo_Custo")),
    .pred_class_0.9 = factor(ifelse(.pred_Alto_Custo >= 0.9,  "Alto_Custo", "Baixo_Custo"), levels = c("Alto_Custo", "Baixo_Custo"))
  )

# Random Forest - Teste
pred_rf_teste = pred_rf_teste %>%
  mutate(
    .pred_class_0.1 = factor(ifelse(.pred_Alto_Custo >= 0.1,  "Alto_Custo", "Baixo_Custo"), levels = c("Alto_Custo", "Baixo_Custo")),
    .pred_class_0.5 = factor(ifelse(.pred_Alto_Custo >= 0.5,  "Alto_Custo", "Baixo_Custo"), levels = c("Alto_Custo", "Baixo_Custo")),
    .pred_class_0.7 = factor(ifelse(.pred_Alto_Custo >= 0.7,  "Alto_Custo", "Baixo_Custo"), levels = c("Alto_Custo", "Baixo_Custo")),
    .pred_class_0.9 = factor(ifelse(.pred_Alto_Custo >= 0.9,  "Alto_Custo", "Baixo_Custo"), levels = c("Alto_Custo", "Baixo_Custo"))
  )

# XGBoost - Teste
pred_xgb_teste = pred_xgb_teste %>%
  mutate(
    .pred_class_0.1 = factor(ifelse(.pred_Alto_Custo >= 0.1,  "Alto_Custo", "Baixo_Custo"), levels = c("Alto_Custo", "Baixo_Custo")),
    .pred_class_0.5 = factor(ifelse(.pred_Alto_Custo >= 0.5,  "Alto_Custo", "Baixo_Custo"), levels = c("Alto_Custo", "Baixo_Custo")),
    .pred_class_0.7 = factor(ifelse(.pred_Alto_Custo >= 0.7,  "Alto_Custo", "Baixo_Custo"), levels = c("Alto_Custo", "Baixo_Custo")),
    .pred_class_0.9 = factor(ifelse(.pred_Alto_Custo >= 0.9,  "Alto_Custo", "Baixo_Custo"), levels = c("Alto_Custo", "Baixo_Custo"))
  )


# ==================================================
# 11. Cálculo de Métricas para Cada Limiar (APENAS PARA OS DADOS DE TREINO)
# ==================================================

metricas_utilizadas = metric_set(precision, recall,f_meas)

metricas_log_treino = bind_rows(
  pred_log_treino %>% metricas_utilizadas(truth = custo_medico, estimate = .pred_class_0.1,event_level = "first") %>% mutate(limiar = 0.1),
  pred_log_treino %>% metricas_utilizadas(truth = custo_medico, estimate = .pred_class_0.5,event_level = "first") %>% mutate(limiar = 0.5),
  pred_log_treino %>% metricas_utilizadas(truth = custo_medico, estimate = .pred_class_0.7,event_level = "first") %>% mutate(limiar = 0.7),
  pred_log_treino %>% metricas_utilizadas(truth = custo_medico, estimate = .pred_class_0.9,event_level = "first") %>% mutate(limiar = 0.9)
) %>% mutate(Modelo = "Regressão Logística")


metricas_rf_treino = bind_rows(
  pred_rf_treino %>% metricas_utilizadas(truth = custo_medico, estimate = .pred_class_0.1,event_level = "first") %>% mutate(limiar = 0.1),
  pred_rf_treino %>% metricas_utilizadas(truth = custo_medico, estimate = .pred_class_0.5,event_level = "first") %>% mutate(limiar = 0.5),
  pred_rf_treino %>% metricas_utilizadas(truth = custo_medico, estimate = .pred_class_0.7,event_level = "first") %>% mutate(limiar = 0.7),
  pred_rf_treino %>% metricas_utilizadas(truth = custo_medico, estimate = .pred_class_0.9,event_level = "first") %>% mutate(limiar = 0.9)
) %>% mutate(Modelo = "Random Forest")

metricas_xgb_treino = bind_rows(
  pred_xgb_treino %>% metricas_utilizadas(truth = custo_medico, estimate = .pred_class_0.1,event_level = "first") %>% mutate(limiar = 0.1),
  pred_xgb_treino %>% metricas_utilizadas(truth = custo_medico, estimate = .pred_class_0.5,event_level = "first") %>% mutate(limiar = 0.5),
  pred_xgb_treino %>% metricas_utilizadas(truth = custo_medico, estimate = .pred_class_0.7,event_level = "first") %>% mutate(limiar = 0.7),
  pred_xgb_treino %>% metricas_utilizadas(truth = custo_medico, estimate = .pred_class_0.9,event_level = "first") %>% mutate(limiar = 0.9)
) %>% mutate(Modelo = "XGBoost")


metricas_log_treino %>% ggplot(aes(x = limiar, y = .estimate , color = .metric,group = .metric)) +
  geom_line(linewidth = 1.2) + # Adiciona as linhas
  geom_point(size = 3)        # Adiciona pontos para destacar os limiares


metricas_rf_treino %>% ggplot(aes(x = limiar, y = .estimate , color = .metric,group = .metric)) +
  geom_line(linewidth = 1.2) + # Adiciona as linhas
  geom_point(size = 3)        # Adiciona pontos para destacar os limiares


metricas_xgb_treino %>% ggplot(aes(x = limiar, y = .estimate , color = .metric,group = .metric)) +
  geom_line(linewidth = 1.2) + # Adiciona as linhas
  geom_point(size = 3)        # Adiciona pontos para destacar os limiares



# ==================================================
# 11.5. Análise Visual de Performance (Dados de Treino)
# ==================================================

matriz_conf_log = pred_log_treino %>%
  conf_mat(truth = custo_medico, estimate = .pred_class_0.5)

autoplot(matriz_conf_log, type = "heatmap") +
  labs(title = "Matriz de Confusão - Regressão Logística (Treino, Limiar 0.5)") +
  scale_fill_gradient(low = "#D6EAF8", high = "#2E86C1")

matriz_conf_rf = pred_rf_treino %>%
  conf_mat(truth = custo_medico, estimate = .pred_class_0.5)

autoplot(matriz_conf_rf, type = "heatmap") +
  labs(title = "Matriz de Confusão - Random Forest (Treino, Limiar 0.5)") +
  scale_fill_gradient(low = "#D6EAF8", high = "#2E86C1")


matriz_conf_xgboost = pred_xgb_treino %>%
  conf_mat(truth = custo_medico, estimate = .pred_class_0.5)

autoplot(matriz_conf_xgboost, type = "heatmap") +
  labs(title = "Matriz de Confusão - XGBOOST (Treino, Limiar 0.5)") +
  scale_fill_gradient(low = "#D6EAF8", high = "#2E86C1")



# ==================================================
# 12. Consolidação e Análise dos Resultados
# ==================================================

# ==================================================
# Calcula métricas para dados de TREINO
# ==================================================
metricas_log_treino = pred_log_treino %>% metricas_utilizadas(truth = custo_medico, estimate = .pred_class_0.5,event_level = "first") %>% mutate(Modelo = "Regressão Logística")
metricas_rf_treino  = pred_rf_treino  %>% metricas_utilizadas(truth = custo_medico, estimate = .pred_class_0.5,event_level = "first") %>% mutate(Modelo = "Random Forest")
metricas_xgb_treino = pred_xgb_treino %>% metricas_utilizadas(truth = custo_medico, estimate = .pred_class_0.5,event_level = "first") %>% mutate(Modelo = "XGBoost")

resultados_treino = bind_rows(
  metricas_log_treino,
  metricas_rf_treino,
  metricas_xgb_treino
)

# ==================================================
# Calcula métricas para dados de TESTE
# ==================================================

metricas_log_teste = pred_log_teste %>% metricas_utilizadas(truth = custo_medico, estimate = .pred_class_0.5,event_level = "first") %>% mutate(Modelo = "Regressão Logística")
metricas_rf_teste  = pred_rf_teste  %>% metricas_utilizadas(truth = custo_medico, estimate = .pred_class_0.5,event_level = "first") %>% mutate(Modelo = "Random Forest")
metricas_xgb_teste = pred_xgb_teste %>% metricas_utilizadas(truth = custo_medico, estimate = .pred_class_0.5,event_level = "first") %>% mutate(Modelo = "XGBoost")

resultados_teste = bind_rows(
  metricas_log_teste,
  metricas_rf_teste,
  metricas_xgb_teste
)

# ==================================================
# Junta e Apresenta Resultados Finais
# ==================================================

# Adiciona coluna indicando etapa
resultados_teste$etapa = "Teste"
resultados_treino$etapa = "Treino"

View(resultados_treino %>% bind_rows(resultados_teste) %>% filter(.metric == "f_meas") %>% arrange(desc(Modelo), desc(etapa) ))


# ==================================================
# 13. Treinar o modelo final e simular produção
# ==================================================

# A regressão logística foi o melhor. Vamos treiná-lo com todos os dados.
modelo_final = fit(final_log_wf, data = df)

# Salvando o modelo
saveRDS(modelo_final, "modelo_final_classificacao.rds")

# --- Simulação de Produção ---
cat("\n\n--- Simulação de Produção ---\n")

# Carregando o modelo salvo
workflow_prod = readRDS("modelo_final_classificacao.rds")

dados_prod = read.csv("dados_producao_classificacao.csv")

# função para automatizar a predição
realizar_predicao = function(workflow, dados){

  previsoes = predict(workflow, new_data = dados, type = "prob") %>%
  bind_cols(dados)
  
  return(previsoes)
}

# Previsões diretas
previsoes = realizar_predicao(workflow_prod,dados_prod)

File: 19-RML/Classwork/Class_03/script_classificacao.py
# ==================================================
# 1. Carregar Pacotes
# ==================================================
import pandas as pd
import numpy as np
import joblib
import matplotlib.pyplot as plt

# Pré-processamento
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.preprocessing import StandardScaler, OneHotEncoder
from sklearn.impute import SimpleImputer
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline

# Modelos a serem utilizados
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from xgboost import XGBClassifier

# Métricas e Visualização
from sklearn.metrics import (
    classification_report,
    f1_score,
    precision_score,
    recall_score
)

# ==================================================
# 2. Carregar e Preparar os Dados
# ==================================================
print("--- 2. Carregando e Preparando os Dados ---")
df = pd.read_csv("aula_02_exemplo_02.csv")
df["custo_medico"] = df["custo_medico"].apply(lambda x : 1 if x == "Alto_Custo" else 0)
X = df.drop("custo_medico", axis=1)
y = df["custo_medico"]
print("Dados carregados e variável alvo transformada.")

# ==================================================
# 3. Divisão dos Dados
# ==================================================
print("\n--- 3. Dividindo os Dados em Treino e Teste ---")
x_treino, x_teste, y_treino, y_teste = train_test_split(
    X, y, test_size=0.2, stratify=y, random_state=42
)
print("Dados divididos: 80% para treino, 20% para teste.")

# ==================================================
# 4. Pré-processamento dos Dados
# ==================================================
print("\n--- 4. Criando e Aplicando o Pipeline de Pré-processamento ---")
variaveis_categoricas = ["sex", "smoker", "region"]
variaveis_numericas = ["age", "bmi", "children"]

etapas_numericas = Pipeline([("imputer", SimpleImputer(strategy="median")), ("scaler", StandardScaler())])
etapas_categoricas = Pipeline([("imputer", SimpleImputer(strategy="most_frequent")), ("encoder", OneHotEncoder(drop="first", handle_unknown="ignore"))])
preprocessador = ColumnTransformer([("num", etapas_numericas, variaveis_numericas), ("cat", etapas_categoricas, variaveis_categoricas)])

x_treino_transformado = preprocessador.fit_transform(x_treino)
x_teste_transformado = preprocessador.transform(x_teste)
print("Pré-processamento concluído.")

# =============================================================================
# 5. Treinamento e Avaliação Individual dos Modelos
# =============================================================================
print("\n--- 5. Treinando e Avaliando os Modelos Individualmente ---")

# --- Regressão Logística ---
print("\n--- Treinando Regressão Logística ---")
modelo_lr = LogisticRegression(max_iter=1000, random_state=42)
param_grid_lr = {'C': [0.1, 1, 10, 100]}
grid_search_lr = GridSearchCV(modelo_lr, param_grid_lr, cv=5, scoring='f1', n_jobs=-1)
grid_search_lr.fit(x_treino_transformado, y_treino)
melhor_modelo_lr = grid_search_lr.best_estimator_

print("\n=== Métricas de Treino (Regressão Logística) ===")
y_pred_treino_lr = melhor_modelo_lr.predict(x_treino_transformado)
print(classification_report(y_treino, y_pred_treino_lr))

# --- Random Forest ---
print("\n--- Treinando Random Forest ---")
modelo_rf = RandomForestClassifier(random_state=42)
param_grid_rf = {'n_estimators': [100, 200], 'max_depth': [5, 10, None], 'min_samples_leaf': [2, 4]}
grid_search_rf = GridSearchCV(modelo_rf, param_grid_rf, cv=5, scoring='f1', n_jobs=-1)
grid_search_rf.fit(x_treino_transformado, y_treino)
melhor_modelo_rf = grid_search_rf.best_estimator_

print("\n=== Métricas de Treino (Random Forest) ===")
y_pred_treino_rf = melhor_modelo_rf.predict(x_treino_transformado)
print(classification_report(y_treino, y_pred_treino_rf))


# --- XGBoost ---
print("\n--- Treinando XGBoost ---")
modelo_xgb = XGBClassifier(random_state=42)
param_grid_xgb = {'n_estimators': [100, 500], 'max_depth': [3, 5], 'learning_rate': [0.01, 0.1]}
grid_search_xgb = GridSearchCV(modelo_xgb, param_grid_xgb, cv=5, scoring='f1', n_jobs=-1)
grid_search_xgb.fit(x_treino_transformado, y_treino)
melhor_modelo_xgb = grid_search_xgb.best_estimator_

print("\n=== Métricas de Treino (XGBoost) ===")
y_pred_treino_xgb = melhor_modelo_xgb.predict(x_treino_transformado)
print(classification_report(y_treino, y_pred_treino_xgb))


# =============================================================================
# 6. Análise de Limiar (Threshold) no Conjunto de TREINO
# =============================================================================
print("\n--- 6. Análise de Limiar para Otimizar Precision vs. Recall ---")

# --- Modelo 1: Regressão Logística ---
print("\n[INFO] Análise de limiar para Regressão Logística (em dados de Treino)...")
y_probas_treino_lr = melhor_modelo_lr.predict_proba(x_treino_transformado)[:, 1]
limiares = np.arange(0.1, 1.0, 0.05)
scores_recall_lr = []
scores_precision_lr = []
for limiar in limiares:
    y_pred_limiar = (y_probas_treino_lr >= limiar).astype(int)
    scores_precision_lr.append(precision_score(y_treino, y_pred_limiar, zero_division=0))
    scores_recall_lr.append(recall_score(y_treino, y_pred_limiar, zero_division=0))
plt.figure(figsize=(10, 6)) 
plt.plot(limiares, scores_precision_lr, label='Precision', marker='o')
plt.plot(limiares, scores_recall_lr, label='Recall', marker='x') 
plt.title('Precision vs. Recall por Limiar - Regressão Logística (Treino)')
plt.xlabel('Limiar de Classificação')
plt.ylabel('Score')
plt.grid(True)
plt.legend()
plt.show()

# --- Modelo 2: Random Forest ---
print("\n[INFO] Análise de limiar para Random Forest (em dados de Treino)...")
y_probas_treino_rf = melhor_modelo_rf.predict_proba(x_treino_transformado)[:, 1]
scores_recall_rf = []
scores_precision_rf = []
for limiar in limiares:
    y_pred_limiar = (y_probas_treino_rf >= limiar).astype(int)
    scores_precision_rf.append(precision_score(y_treino, y_pred_limiar, zero_division=0))
    scores_recall_rf.append(recall_score(y_treino, y_pred_limiar, zero_division=0))
plt.figure(figsize=(10, 6))
plt.plot(limiares, scores_precision_rf, label='Precision', marker='o')
plt.plot(limiares, scores_recall_rf, label='Recall', marker='x')
plt.title('Precision vs. Recall por Limiar - Random Forest (Treino)')
plt.xlabel('Limiar de Classificação')
plt.ylabel('Score')
plt.grid(True)
plt.legend()
plt.show()

# --- Modelo 3: XGBoost ---
print("\n[INFO] Análise de limiar para XGBoost (em dados de Treino)...")
y_probas_treino_xgb = melhor_modelo_xgb.predict_proba(x_treino_transformado)[:, 1]
scores_recall_xgb = []
scores_precision_xgb = []
for limiar in limiares:
    y_pred_limiar = (y_probas_treino_xgb >= limiar).astype(int)
    scores_precision_xgb.append(precision_score(y_treino, y_pred_limiar, zero_division=0))
    scores_recall_xgb.append(recall_score(y_treino, y_pred_limiar, zero_division=0))
plt.figure(figsize=(10, 6))
plt.plot(limiares, scores_precision_xgb, label='Precision', marker='o')
plt.plot(limiares, scores_recall_xgb, label='Recall', marker='x')
plt.title('Precision vs. Recall por Limiar - XGBoost (Treino)')
plt.xlabel('Limiar de Classificação')
plt.ylabel('Score')
plt.grid(True)
plt.legend()
plt.show()

# ==================================================
# 7. Consolidação e Análise Comparativa dos Resultados
# ==================================================
y_pred_teste_xgb = melhor_modelo_xgb.predict(x_teste_transformado)
y_pred_teste_rf = melhor_modelo_rf.predict(x_teste_transformado)
y_pred_teste_lr = melhor_modelo_lr.predict(x_teste_transformado)

print("\n--- 7. Tabela Comparativa de Performance (F1-Score) ---")
results_list = []
results_list.append({'Modelo': 'Regressão Logística', 'Etapa': 'Treino', 'F1_Score': f1_score(y_treino, y_pred_treino_lr)})
results_list.append({'Modelo': 'Regressão Logística', 'Etapa': 'Teste', 'F1_Score': f1_score(y_teste, y_pred_teste_lr)})
results_list.append({'Modelo': 'Random Forest', 'Etapa': 'Treino', 'F1_Score': f1_score(y_treino, y_pred_treino_rf)})
results_list.append({'Modelo': 'Random Forest', 'Etapa': 'Teste', 'F1_Score': f1_score(y_teste, y_pred_teste_rf)})
results_list.append({'Modelo': 'XGBoost', 'Etapa': 'Treino', 'F1_Score': f1_score(y_treino, y_pred_treino_xgb)})
results_list.append({'Modelo': 'XGBoost', 'Etapa': 'Teste', 'F1_Score': f1_score(y_teste, y_pred_teste_xgb)})

results_df = pd.DataFrame(results_list)
results_df_sorted = results_df.sort_values(by=['Modelo', 'Etapa'])
print(results_df_sorted)

# ==================================================
# 8. Treinamento do Modelo Final e Simulação de Produção
# ==================================================
print("\n--- 8. Preparando o Modelo Final para Produção ---")
print("Com base nos relatórios de classificação e na análise de limiar, o LR foi escolhido.")
modelo_final = Pipeline([('preprocessador', preprocessador), ('modelo', melhor_modelo_lr)])
modelo_final.fit(X, y)
joblib.dump(modelo_final, 'modelo_final_classificacao.joblib')
print("\nModelo final salvo como 'modelo_final_classificacao.joblib'")

print("\n--- Simulação de Uso em Produção ---")
modelo_producao = joblib.load('modelo_final_classificacao.joblib')
dados_novos = pd.read_csv("dados_producao_classificacao.csv")
predicoes_prod = modelo_producao.predict(dados_novos)

dados_novos['pred'] = predicoes_prod
print(dados_novos)